Simplify and clarify code for local-cluster

abailly-iohk opened the pull request:
The local-cluster project was the first code we wrote and some stuff in it are redundant with stuff we wrote later, or ill-named.

* Replace local-cluster Logging with Hydra.Logging
* Move Ports utility module to Hydra.Network and use its feature there, namely in the `NetworkSpec`  tests
* Rename Lib -> CardanoCluster and Node -> CardanoNode

The changes introduced made it even more apparent how flaky our `NetworkSpec` tests were so this PR also fixes that by:

* Having a simpler way to check we broadcast to all nodes in the tests
* Increasing timeout of Ouroboros network tests to 30s to give time to the framework to retry failed connection attempts

abailly committed:
Simplify and clarify code for local-cluster

The local-cluster project was the first code we wrote and some stuff
in it are redundant with stuff we wrote later, or ill-named.

* Replace local-cluster Logging with Hydra.Logging
* Move Ports utility module to Hydra.Network and use its feature there
* Rename Lib -> CardanoCluster and Node -> CardanoNode

ghost commented:
This PR was triggered by this ticket in our board's Red Bin: https://miro.com/app/board/o9J_lRyWcSY=/?moveToWidget=3074457360130079706&cot=14

abailly committed:
Allocated random ports for Network tests

This does not seem to improve the stability of the tests but it's
still better to not have hard coded ports like that lying around.

KtorZ commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -40,8 +41,9 @@ spec = describe "Networking layer" $ do
       showLogsOnFailure $ \tracer -> failAfter 10 $ do
-        withOuroborosNetwork tracer (lo, 45778) [(lo, 45779)] (const @_ @(HydraMessage SimpleTx) $ pure ()) $ \hn1 ->
-          withOuroborosNetwork @(HydraMessage SimpleTx) tracer (lo, 45779) [(lo, 45778)] (putMVar received) $ \_ -> do
+        [port1, port2] <- fmap fromIntegral <$> randomUnusedTCPPorts 2
```
:clap:

KtorZ replied:
Little remark maybe, the `fmap fromIntegral` is for converting from a Port type to Int or vice-versa? Would it make sense to align which ever side is using the untyped version?

KtorZ replied:
>  I think it's better to have randomUnusedTCPPorts produce PortNumber. WDYT?

Yup!

abailly committed:
Reverse accumulated logs on error

abailly committed:
Use same retry logic for test with only 2 nodes

KtorZ commented on hydra-node/src/Hydra/Logging.hs:
```diff
@@ -92,3 +95,12 @@ transformLogObject transform tr = Tracer $ \a -> do
+showLogsOnFailure action = do
+  tvar <- newTVarIO []
+  action (traceInTVar tvar)
+    `onException` (atomically (readTVar tvar) >>= mapM_ (say . show) . reverse)
```
:+1:

abailly committed:
Remove dependency on Hydra.Logging in local-cluster

abailly committed:
Move Ports module to hydra-prelude

abailly committed:
Increase timeout of Network tests to 30s

There is a fixed retry window of 10s in Ouroboros networking code that
cannot be configured hence having a 10s timeout runs a high risk of
failing because a node which fails to connect won't have the time to
reconnect.

abailly committed:
Fix compilation error

KtorZ commented on hydra-node/hydra-node.cabal:
```diff
@@ -70,6 +70,7 @@ common project-config
     -Wincomplete-uni-patterns
     -Wpartial-fields
     -Wredundant-constraints
+    -Wunused-packages
```
I thought we said we wanted to remove that one because it was making ghci buggy :thinking: ?

KtorZ commented on hydra-node/src/Hydra/Network/Ouroboros.hs:
```diff
@@ -192,10 +194,10 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
+        { nstMuxTracer = contramap (TraceMux localHost) tracer
+        , nstHandshakeTracer = contramap (TraceHandshake localHost) tracer
+        , nstErrorPolicyTracer = contramap (TraceErrorPolicy localHost) tracer
+        , nstAcceptPolicyTracer = contramap (TraceAcceptPolicy localHost) tracer
```
In the ouroboros-network codebase, I've often seen the `WithXXX` pattern for encapsulating traces in a certain context, while avoid to repeat the context on each branch of the trace, like:

```hs
data WithHost trace = WithHost Host trace
```

It keeps the actual trace (here `TraceOuroborosNetwork`) clean, while still providing a nice context to the tracer.

KtorZ commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -34,61 +34,78 @@ import Test.QuickCheck (
+        [port1, port2] <- fmap fromIntegral <$> randomUnusedTCPPorts 2
+        withOuroborosNetwork tracer (lo, port1) [(lo, port2)] (const @_ @Integer $ pure ()) $ \hn1 ->
+          withOuroborosNetwork @Integer tracer (lo, port2) [(lo, port1)] (atomically . writeTQueue received) $ \_ -> do
+            withAsync (1 `broadcastFrom` hn1) $ \_ ->
```
Why the async here :thinking: ?

KtorZ commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -34,61 +34,78 @@ import Test.QuickCheck (
+  Network IO Integer ->
+  TQueue IO Integer ->
+  TQueue IO Integer ->
+  TQueue IO Integer ->
```
:sweat_smile: ... nice signature... I guess we could generalize it to `n` nodes although it's not strictly needed and work as is.

abailly committed:
Address review Comments

abailly committed:
Silly typo repeating port number

abailly committed:
Fix compilation errors following rebase

abailly committed:
Merge branch 'master' into abailly-iohk/unify-local-cluster-code

ch1bo commented on hydra-node/src/Hydra/Network/Ouroboros.hs:
```diff
@@ -254,9 +256,14 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
+data TraceOuroborosNetwork msg
   = TraceSubscriptions (WithIPList (SubscriptionTrace SockAddr))
   | TraceErrorPolicy (WithAddr SockAddr ErrorPolicyTrace)
   | TraceAcceptPolicy AcceptConnectionsPolicyTrace
```
Seeing the naming convention of ouroboros-network of likely being `XXXTrace` for trace types. Shall we adopt that as well? We have at least `TraceXXX` and `XXXLog` in use.

ch1bo replied:
No change required here, just starting the conversation :)

ch1bo commented on hydra-node/src/Hydra/Network/Ouroboros.hs:
```diff
@@ -254,9 +256,14 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
       }
 
-data TraceOuroborosNetwork
+data WithHost trace = WithHost Host trace
```
Isn't hlint warning that this could be a `newtype`? if yes, I'd follow that suggestion

ch1bo commented on hydra-node/src/Hydra/Network/ZeroMQ.hs:
```diff
@@ -27,9 +27,9 @@ import System.ZMQ4.Monadic (
+  | MessageSent Host LByteString
+  | LogMessageReceived Host Text
+  | SubscribedTo Host [String]
   deriving (Show)
```
Could: use the `WithHost` type you introduced

ch1bo commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -34,61 +34,84 @@ import Test.QuickCheck (
-  tryBroadcast = do
-    broadcast network requestTx
-    forM_ receivers $ \var -> failAfter 1 $ atomically (takeTMVar var) `shouldReturn` requestTx
+  parallelBroadcast :: IO () -> [(PortNumber, Network IO Integer, TQueue IO Integer)] -> IO ()
+  parallelBroadcast check [] = check
+  parallelBroadcast check ((port, network, _) : rest) =
+    withAsync (fromIntegral port `broadcastFrom` network) $ \_ -> parallelBroadcast check rest
```
Could: This function is passing `check` around to just have it as "core" to the `withAsync` chain.. an alternative (which reads a bit clearer to me, would be to `race_ (parallelBroadcast networks) checkQueues` above and consequently:
```suggestion
  parallelBroadcast :: [(PortNumber, Network IO Integer, TQueue IO Integer)] -> IO ()
  parallelBroadcast [] = pure ()
  parallelBroadcast [(port, network, _)] = fromIntegral port `broadcastFrom` network
  parallelBroadcast ((port, network, _) : rest) =
    withAsync (fromIntegral port `broadcastFrom` network) $ \_ -> parallelBroadcast rest
```

ch1bo commented on local-cluster/local-cluster.cabal:
```diff
@@ -84,6 +82,7 @@ library
     , hspec-expectations
     , http-conduit
     , hydra-prelude
+    , hydra-node
```
As we have this here anyway, I guess I'd prefer the Hydra.Network.Ports to be in `hydra-node` instead of the `hydra-prelude`. May I move it or WDYT?

ch1bo reviewed (approved):
Only minor things remain, but can also be merged like this

ghost commented:
@ch1bo All valid points, I will address them after the merge.
